pkg/solana/chainreader/bindings.go

```diff
@@ -13,167 +14,420 @@ import (
+	if !sharesAddress {
+		return nil, false
+	}
+
+	return shareGroup, sharesAddress
 }
 
-func (b *bindingsRegistry) handleAddressSharing(boundContract *types.BoundContract) error {
-	shareGroup, isInAGroup := b.getShareGroup(boundContract.Name)
+func (r *bindingsRegistry) handleAddressSharing(boundContract *types.BoundContract) error {
+	shareGroup, isInAGroup := r.getShareGroup(boundContract.Name)
 	if !isInAGroup {
 		return nil
 	}
 
-	shareGroup.mux.Lock()
-	defer shareGroup.mux.Unlock()
-
 	// set shared address to the binding address
-	if shareGroup.address.IsZero() {
+	if shareGroup.getAddress().IsZero() {
 		key, err := solana.PublicKeyFromBase58(boundContract.Address)
 		if err != nil {
 			return err
 		}
-		b.addressShareGroups[boundContract.Name].address, shareGroup.address = key, key
-	} else if boundContract.Address != shareGroup.address.String() && boundContract.Address != "" {
-		return fmt.Errorf("namespace: %q shares address: %q with namespaceBindings: %v and cannot be bound with a new address: %s", boundContract.Name, shareGroup.address, shareGroup.group, boundContract.Address)
+
+		shareGroup.setAddress(key)
+	} else if boundContract.Address != shareGroup.getAddress().String() && boundContract.Address != "" {
+		return fmt.Errorf("namespace: %q shares address: %q with namespaceBindings: %v and cannot be bound with a new address: %s", boundContract.Name, shareGroup.getAddress(), shareGroup.group, boundContract.Address)
+	}
+
+	// side-effect of updating the bound contract address to group-shared address
+	boundContract.Address = shareGroup.getAddress().String()
+
+	return nil
+}
+
+type namespaceBinding struct {
+	// static data
+	name string
+
+	// dynamic thread-safe data
+	mu      sync.RWMutex
+	readers map[string]readBinding
+	bound   map[solana.PublicKey]bool
+}
+
+func newNamespaceBinding(namespace string) *namespaceBinding {
+	return &namespaceBinding{
+		name:    namespace,
+		readers: make(map[string]readBinding),
+		bound:   make(map[solana.PublicKey]bool),
 	}
+}
+
+func (b *namespaceBinding) SetCodecs(codec types.RemoteCodec) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, rb := range b.readers {
+		rb.SetCodec(codec)
+	}
+}
+
+func (b *namespaceBinding) SetModifiers(modifier commoncodec.Modifier) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	for _, rb := range b.readers {
+		rb.SetModifier(modifier)
+	}
+}
+
+func (b *namespaceBinding) Bind(ctx context.Context, reg filterRegistrar, address solana.PublicKey) error {
+	if b.bindingExists(address) {
+		return nil
+	}
+
+	b.setBinding(address)
 
-	boundContract.Address = shareGroup.address.String()
 	return nil
 }
 
-func (b *bindingsRegistry) getShareGroup(nameSpace string) (*addressShareGroup, bool) {
-	shareGroup, sharesAddress := b.addressShareGroups[nameSpace]
-	if !sharesAddress {
-		return nil, false
+func (b *namespaceBinding) BindReaders(ctx context.Context, address solana.PublicKey) error {
+	b.mu.RLock()
```

Comment:
And also here for write locks?

Reply:
The `*namespaceBinding` is only doing a read on `b.readers`.